rust/src/database/managed_trie/store.rs

```diff
@@ -0,0 +1,511 @@
+// Copyright (c) 2025 Sonic Operations Ltd
+//
+// Use of this software is governed by the Business Source License included
+// in the LICENSE file and at soniclabs.com/bsl11.
+//
+// Change Date: 2028-4-16
+//
+// On the date above, in accordance with the Business Source License, use of
+// this software will be governed by the GNU Lesser General Public License v3.
+
+use std::sync::RwLockWriteGuard;
+
+use crate::{
+    database::managed_trie::{
+        TrieCommitment, TrieUpdateLog,
+        managed_trie_node::{StoreAction, UnionManagedTrieNode},
+    },
+    error::{BTResult, Error},
+    node_manager::NodeManager,
+    types::{Key, Value},
+};
+
+/// Stores the given key-value pair into the managed trie rooted at `root_id`.
+///
+/// In case the root node of the trie changes, the `root_id` guard is updated accordingly.
+///
+/// At most two nodes are write-locked at any given time during tree traversal, allowing for
+/// concurrent lookup/store operations on different parts of the tree. The lock on the root ID is
+/// held until the algorithm has descended two levels into the tree.
+///
+/// The `update_log` is updated to reflect which nodes need to have their commitments recomputed
+/// after the store operation.
+#[cfg_attr(not(test), expect(unused))]
+pub fn store<T: UnionManagedTrieNode>(
+    root_id: RwLockWriteGuard<T::Id>,
+    key: &Key,
+    value: &Value,
+    manager: &impl NodeManager<Id = T::Id, NodeType = T>,
+    update_log: &TrieUpdateLog<T::Id>,
+) -> BTResult<(), Error>
+where
+    T::Id: Copy + Eq + std::hash::Hash + std::fmt::Debug,
+{
+    let mut parent_lock = None;
+    // Wrap the root ID lock into an Option so we can release it once we are deep enough in the tree
+    let mut root_id = Some(root_id);
+    let mut current_id = **root_id.as_ref().unwrap();
+    let mut current_lock = manager.get_write_access(current_id)?;
+    let mut depth = 0;
+
+    loop {
+        match current_lock.next_store_action(key, depth, current_id)? {
+            StoreAction::Store { index } => {
+                let prev_value = current_lock.store(key, value)?;
+
+                let mut trie_commitment = current_lock.get_commitment();
+                trie_commitment.store(index, prev_value);
+                current_lock.set_commitment(trie_commitment)?;
+                update_log.mark_dirty(depth as usize, current_id);
+
+                return Ok(());
+            }
+            StoreAction::Descend { index, id } => {
+                let mut trie_commitment = current_lock.get_commitment();
+                trie_commitment.modify_child(index);
+                current_lock.set_commitment(trie_commitment)?;
```

Comment:
Would it be possible to hand out a mutable reference to the commitment and modify it in place?

Reply:
The problem is that I need to access the commitment through both mutable and immutable references. So I'd need a `get_commitment` and `get_commitment_mut`. Not sure if that is better :shrug: